### DISCUSSION TITLE
Deeplinking in iOS

### DISCUSSION THREAD

**Discussion by jmgoree** [arkivanov/Decompose#784] [Q&A]
Are there any examples of passing an iOS universal link to decompose. I have successfully implemented on the Android side, however need some assistance/ideas on how to accomplish it on the iOS side. 

I am able to capture the url in the swift code here:
`@main
struct iOSApp: App {
    
    let Logger: DatadogLogger
    @StateObject private var deeplink = Deeplink()
    @UIApplicationDelegateAdaptor(AppDelegate.self) var appDelegate

    init() {
        self.Logger = DatadogLogger()
    }

	var body: some Scene {
		WindowGroup {
            ContentView()
                .environmentObject(deeplink)
                .onOpenURL(perform: { url in
                    print ("before " + String(deeplink.publicUserId))
                    deeplink.publicUserId = (!url.absoluteString.isEmpty ? Int(url.lastPathComponent) : -1) ?? -1
                    print ("after " + String(deeplink.publicUserId))
            })
		}
	}
}`

However the problem i have is passing that data to the compose code to recreate/update the mainviewcontroller. Is there an iOS equivalent of the android handleDeeplink?

**Comment by arkivanov** [comment]
Please take a look at [Confetti](https://github.com/arkivanov/Confetti), it has an [example](https://github.com/arkivanov/Confetti/blob/98e2ec256dfedb972620c084b290c6d348370608/iosApp/iosApp/iOSApp.swift#L16) of deeplinking on iOS.

**Reply by jmgoree** [reply]
Thanks for the example! I incorporated it into my code. I can capture the deeplink in logs, but having an issue forcing the compose controller to update. I think I may be missing something pretty simple

``` 
class RootHolder : ObservableObject {
    var lifecycle: LifecycleRegistry
    var root: DefaultDraftCardComponent

    init() {
        lifecycle = LifecycleRegistryKt.LifecycleRegistry()

        root = DefaultDraftCardComponent(
            componentContext: DefaultComponentContext(lifecycle: lifecycle),
            initialPublicUserId: -1
        )

        LifecycleRegistryExtKt.create(lifecycle)
    }
    
    func onPublicUserProfileDeepLink(userId: Int) {
        lifecycle.destroy()
        lifecycle = LifecycleRegistryKt.LifecycleRegistry()
        root = DefaultDraftCardComponent(
            componentContext: DefaultComponentContext(lifecycle: lifecycle),
            initialPublicUserId: userId as? KotlinInt
        )
    }

    deinit {
        // Destroy the root component before it is deallocated
        LifecycleRegistryExtKt.destroy(lifecycle)
    }

@main
struct iOSApp: App {
    
    let Logger: DatadogLogger
    @UIApplicationDelegateAdaptor(AppDelegate.self) var appDelegate
    
    @Environment(\.scenePhase)
    var scenePhase: ScenePhase
    
    var rootHolder: RootHolder { appDelegate.rootHolder }
    

    init() {
        self.Logger = DatadogLogger()
    }

	var body: some Scene {
		WindowGroup {
            ContentView(rootHolder: rootHolder)
                .onOpenURL(perform: { url in
                    let publicUserId = (!url.absoluteString.isEmpty ? Int(url.lastPathComponent) : -1) ?? -1
                    rootHolder.onPublicUserProfileDeepLink(userId: publicUserId)
                    print ("Deeplink recieved " + String(publicUserId))
            }).onChange(of: scenePhase) { newPhase in
                switch newPhase {
                case .background: LifecycleRegistryExtKt.stop(rootHolder.lifecycle)
                case .inactive: LifecycleRegistryExtKt.pause(rootHolder.lifecycle)
                case .active: LifecycleRegistryExtKt.resume(rootHolder.lifecycle)
                @unknown default: break
                }
            }
		}
	}
}

struct ComposeView: UIViewControllerRepresentable {
    let root: DefaultDraftCardComponent

    func makeUIViewController(context: Context) -> UIViewController {
        MainViewControllerKt.MainViewController(
            adBannerView: { () -> UIViewController in
                            let adBannerView = VStack {
                                AdViewBanner()
                            }
                            return UIHostingController(rootView: adBannerView)
                        },
            rootComponent: root
        )
    }

    func updateUIViewController(_ uiViewController: UIViewController, context: Context) {
        MainViewControllerKt.MainViewController(
            adBannerView: { () -> UIViewController in
                            let adBannerView = VStack {
                                AdViewBanner()
                            }
                            return UIHostingController(rootView: adBannerView)
                        },
            rootComponent: root
        )
    }
}

struct ContentView: View {
    @ObservedObject
    var rootHolder: RootHolder
    
    var body: some View {
        ComposeView(root: rootHolder.root)
            .ignoresSafeArea(.keyboard) // Compose has own keyboard handler
    }
}
}

```


Within the iosmain section of the shared section

```
fun MainViewController(adBannerView: () -> UIViewController, rootComponent: DefaultDraftCardComponent) = ComposeUIViewController {
    val swiftUiBannerAd = @Composable {
        UIKitViewController(
            // If you don't set 50.dp height it occurs error
            modifier = Modifier
                .fillMaxWidth()
                .height(50.dp),
            factory = adBannerView
        )
    }

    val koinApplication = koinApplication { modules(
        sessionModule,
        datastoreModule,
        databaseModule,
        userQueriesModule,
        sportQueriesModule,
        schoolQueriesModule,
        statQueriesModule,
        positionQueriesModule,) }

    Application().DraftCardApplication(
        application = koinApplication,
        rootComponent = rootComponent,
        adBannerView = swiftUiBannerAd,
        logger = IOSLogger())
}

```

**Reply by arkivanov** [reply]
I think you can try [annotating](https://github.com/arkivanov/Confetti/blob/98e2ec256dfedb972620c084b290c6d348370608/iosApp/iosApp/AppDelegate.swift#L15) your root component with `@Published` and [replace](https://github.com/arkivanov/Confetti/blob/98e2ec256dfedb972620c084b290c6d348370608/iosApp/iosApp/AppDelegate.swift#L30) the component with a new instance when you receive a deep link (which seems like you are already doing).

**Reply by jmgoree** [reply]
Thanks Again for the help! After debugging a bit more, if i dont use compose in the shared code to drive the UI. Deeplinking works. However once i use that path from iOS the changes/new instance doesnt update the compose code. I need to investigate possibly building a shared flow or similar that the compose side could listen to and change accordingly. But that seems like recreating what the @Published and new instance creation already should do

**Reply by arkivanov** [reply]
I didn't try yet, but you may need to wrap the component with MutableState somehow. I could try later if that didn't work

**Reply by jmgoree** [reply]
Interesting, I'll play around with it but if you happen to come across any examples I'd greatly appreciate it. Again thanks for the help

**Reply by arkivanov** [reply]
After thinking more about it, since you are using Compose, I think you don't need `@Published` for the root component. You can try creating the following Kotlin class next to your root `MainViewController`. Then try using it in your Swift app delegate instead of `@Published`, and call `set` when a instance of the root component is created. Pass the holder to your `MainViewController` and use the component from there. This should make your composable recomposing on component change.

```kotlin
class RootComponentHolder(component: RootComponent) {
    internal val component = mutableStateOf(component)

    fun set(component: RootComponent) {
        this.component.value = component
    }
}
```

**Reply by jmgoree** [reply]
@arkivanov Thanks that worked. I ended up taking it a step further and letting that "holder" class notify the app as a whole via a shared flow to gracefully handle new info provided from the deeplinks/anything external to compose.
